hostmetrics: use WMI to fetch ppid

braydonk opened the pull request:
**Description:** <Describe what has changed.>
Scraping process metrics on Windows only really has one truly expensive operation: getting the Parent Process ID. gopsutil does this with the [CreateToolhelp32Snapshot](https://learn.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot) Win32 API call. This takes a snapshot of a large amount of process details and ends up being very expensive. This is causing CPU performance issues for Windows users.

This PR does the following in an attempt to address this performance issue:
* Repurpose the handlecount package to be for querying any Window Management Interface process data and add ParentProcessID to that querying process
* Change the default ParentProcessID fetching process to be through the new WMI fetching API, with a new configuration option `disable_wmi` to turn WMI off in the rare case where you can't use WMI
* Whether using WMI or not, this PR also adds a guard around fetching the parent process ID that will only check it if the `process.ppid` resource attribute is enabled. This means users in very resource constrained environments can opt out of fetching the metric regardless.

**Link to tracking Issue:** #32947

**Testing:**
I tested this by doing two builds of the collector and comparing the performance against each other. With the WMI feature flag enabled, there is a reduction in the CPU usage. (red = before, green = after)
![perfmon screenshot showing the collector without the feature flag using on average 40% more CPU](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/93549768/65df47ba-3ba7-4818-8aa1-d654419c8ba4)

I also compared the original way of getting parent process ID against disabling the resource attribute. The difference is even more drastic. (red = before, green = after)
![perfmon screenshot showing the collector that has the parent process ID resource attribute labelled using drastically less CPU](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/93549768/47eb02d2-97c6-4e77-8426-40832135a661)

**Documentation:**
Added the new configuration value to the README. Added documentation note to `process.handles` to note that it has a hard requirement on WMI.

braydonk committed:
hostmetrics: add WMI querying ppid featureflag

Scraping metrics on Windows largely has one truly expensive operation:
getting the Parent Process ID. gopsutil does this with the
[CreateToolhelp32Snapshot](https://learn.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot)
Win32 API call. This takes a snapshot of a large amount of process
details and ends up being very expensive. This is causing CPU
performance issues for Windows users.

This PR does the following in an attempt to address this perforamnce
issue:
* Repurpose the handlecount package to be for querying any WMI process
  data and add ParentProcessID to that querying process
* Add the featureflag `hostmetrics.process.wmiParentProcessID` which
  will enable using the WMI query for parent process ID fetching
* Whether the featureflag is on or off, this PR also adds a guard around
  fetching the parent process ID that will only check it if the resource
  attribute for ppid is enabled. This means users in very resource
  constrained environments can opt out of that fetching either way

braydonk committed:
fix lint errors

braydonk committed:
fix more lint errors

braydonk committed:
change WMI usage to a config field

braydonk committed:
adjust the changelog to reflect new changes

braydonk committed:
add documentation

braydonk committed:
generate documentation changes

braydonk commented:
I finally had some time to rework this PR with new changes since the old one was closed when I let it go stale. Hopefully this time we can get it pushed through.

braydonk committed:
fix testifylint error

braydonk commented:
Not stale

braydonk commented:
Not stale

andrzej-stencel commented on receiver/hostmetricsreceiver/README.md:
```diff
@@ -129,6 +130,7 @@ The following settings are optional:
 - `mute_process_cgroup_error` (default: false): mute the error encountered when trying to read the cgroup of a process the collector does not have permission to read
 - `mute_process_exe_error` (default: false): mute the error encountered when trying to read the executable path of a process the collector does not have permission to read (Linux only)
 - `mute_process_user_error` (default: false): mute the error encountered when trying to read a uid which doesn't exist on the system, eg. is owned by a user that only exists in a container.
+- `disable_wmi` (default: false) disable the use of Windows Management Interface for fetching process information. Please note that certain metrics cannot be collected without WMI disabled.
```
I propose `enable_wmi` (default `true`) or maybe even better `wmi_enabled` to avoid double negation with `disable_wmi: false`.

pjanotti replied:
I'm not sure that it is a good idea to tie the data retrieved with the underlying technology to get it. Can't we instead rely on logging proper warning at startup if WMI is not available and use the enable/disable metrics to allow the user to suppress the warnings?

braydonk replied:
`wmi_enabled` change done in [acb81f8](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35337/commits/acb81f8b2a5b1c953366203f442dc46f33ce9c4f)

Responding to Paulo's comment separately but resolving this specific thread

andrzej-stencel commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:
```diff
@@ -466,3 +489,24 @@ func (s *scraper) scrapeAndAppendSignalsPendingMetric(ctx context.Context, now p
+func configureWindowsSettings(s *scraper) error {
+	if s.config.DisableWMI {
+		if s.config.Metrics.ProcessHandles.Enabled {
+			return errProcessHandlesRequiresWMI
```
This should be checked in `Config::Validate` to fail early and have the collector's `validate` command use it.

braydonk replied:
Done in [acb81f8](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35337/commits/acb81f8b2a5b1c953366203f442dc46f33ce9c4f)

andrzej-stencel commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/internal/wmiprocinfo/manager_windows.go:
```diff
@@ -0,0 +1,101 @@
+func (wmiProcessQueryer) wmiProcessQuery() (map[int64]*wmiProcInfo, error) {
+	processes := []Win32_Process{}
+	// Based on reflection of Win32_Process type, this creates the following query:
+	// `get-wmiobject -query "select ProcessId, ParentProcessId, HandleCount from Win32_Process"`
```
This method retrieves both parent process ID and process handle count on every run, regardless of whether both are needed.

Would it make sense to call different commands depending on which information is needed? Would it help in performance to for example only retrieve parent process ID and not retrieve handle count, assuming the user didn't enable the `process.handles` metric?

The default configuration for the process scraper is to require parent process ID (as the `process.parent_pid` resource attribute is enabled by default) and not require handle count (as the `process.handles` metric is disabled by default).

braydonk replied:
Accomplished via manual query construction based on the passed in config in [acb81f8](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35337/commits/acb81f8b2a5b1c953366203f442dc46f33ce9c4f)

braydonk commented:
Thanks for the reviews @pjanotti and @andrzej-stencel, I'm acknowledging that I've seen them but can't address yet. Right after I asked for that review at the Collector SIG I had to be out for a bit so I wasn't able to get to it. Thanks for getting to it promptly even though I ended up being unable to answer. Will respond next week!

pjanotti commented:
@braydonk let's know if you would like some help to move this forward, it is a good issue to improve.

braydonk committed:
merge main branch

braydonk committed:
address review comments, increase test coverage

braydonk committed:
Merge branch 'main' into ppid_wmi

braydonk commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/documentation.md:
```diff
@@ -116,7 +116,7 @@ Number of disk operations performed by the process.
 Number of handles held by the process.
 
-This metric is only available on Windows.
+This metric is only available on Windows. It requires Windows Management Interface to be enabled.
```
I don't remember which commit exactly, hopefully this one cause the merge commit is impossible to load on GitHub [acb81f8](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35337/commits/acb81f8b2a5b1c953366203f442dc46f33ce9c4f)

But it is fixed.

braydonk commented on receiver/hostmetricsreceiver/internal/scraper/processscraper/internal/wmiprocinfo/manager_windows.go:
```diff
@@ -0,0 +1,101 @@
+	processes := []Win32_Process{}
+	// Based on reflection of Win32_Process type, this creates the following query:
+	// `get-wmiobject -query "select ProcessId, ParentProcessId, HandleCount from Win32_Process"`
+	q := wmi.CreateQuery(&processes, "")
```
Done in [acb81f8](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35337/commits/acb81f8b2a5b1c953366203f442dc46f33ce9c4f)

I don't directly expect much impact to allowing missing fields based on a read of the code; looks like the only codepath difference when allowing missing fields is for a particular error to be ignored here: https://github.com/yusufpapurcu/wmi/blob/master/wmi.go#L354

I will try and do some due dilligence on the effect of including/not include certain struct fields one way or another. When I originally tested the difference between always fetching both fields vs only fetching one, I did not notice a substantial difference in WMI's performance. However this test was done a while ago, and it would be a good idea to do it again.

braydonk commented:
On Oct 28, 2024 @pjanotti wrote:
```
I'm not sure that it is a good idea to tie the data retrieved with the underlying technology to get it. Can't we instead rely on logging proper warning at startup if WMI is not available and use the enable/disable metrics to allow the user to suppress the warnings?
```

For this initial revival of the PR I stuck with the spirit of my original approach, this was my thought process.

I agree that it is odd at a surface level to seemingly arbitrarily relate certain metrics and the retrieval method. The reason I think it's a good idea is because the retrieval method has an unfortunate underlying nuance, that being the dependence on a separate service. I think it's very rare for WMI to be completely inaccessible on a given machine, but a user may still want to disallow the collector to use the service.

That's why I made it so the flow is something like this:
* There is a config option to completely disallow WMI in the case where a user wants that, it is on by default
* If it is off and `process.handles` is enabled, config validation fails because that is the only possible method to collect that metric
* If it is off and `process.parent_pid` is enabled (which it is by default), it will fall back to the original process snapshotting method of retrieving the attribute
* If WMI is allowed but `process.handles` and `process.parent_pid` are both disabled, no WMI interaction will take place

This all hinges on the premise that the user should be allowed to unilaterally allow/disallow the collector to use WMI. If this isn't actually as important as I'm making it out to be, we can reconsider.

braydonk committed:
fix build for non-windows systems

pjanotti reviewed (commented):
This is already a performance improvement over the current code. However, looking a bit at how the process data is retrieved on Windows it becomes clear that the optimization instead could be done at the time that the PIDs are captured `handles, err := s.getProcessHandles(ctx)`. Internally, `gopsutils`, uses `EnumProcesses` to get the list of PIDs, then one by one we execute code to get other information about each process. So it calls `CreateToolhelp32Snapshot` for each process to get the parent PID, instead it should call `CreateToolhelp32Snapshot` instead of `EnumProcesses` when obtaining the initial list of PIDs. Besides performance the current code has issues regarding correctness and availability of information (processes can terminate between the calls and PIDs can be reused). However, this is a more involved optimization than the current PR since we may need to call Win32 APIs directly instead of relying on `gopsutils`.

Given the current implementation I wonder if using WMI to get all properties would actually perform better. The marshalling cost of all requested properties, for a high enough number of processes, can make WMI more expensive than the Win32 API, however, given the current implementation I suspect WMI can perform better - anyway, needs to be measured.

pjanotti commented:
@braydonk 
> The reason I think it's a good idea is because the retrieval method has an unfortunate underlying nuance, that being the dependence on a separate service.

Given these nuances and [my comment about the current code](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35337#pullrequestreview-2637749550). Unless there is some property that we think is not available in WMI we should consider an option like `use_wmi` that does a single WMI query for all properties, when `use_wmi` is disabled we could do the Win32 API route and in that case some of the info may not be available (parent PID can be made cheaply available without WMI, not sure about handle count).

braydonk commented:
Thanks for the review @pjanotti!

First, updated performance data. Here's a perfmon screenshot running two collectors next to each other.

![image](https://github.com/user-attachments/assets/e3d4b9ca-a369-4764-8478-5a28687f2aea)

(It's actually the same binary, but I made two copies with different names for screenshot demonstration)

WMI Off: 
[config file](https://gist.github.com/braydonk/9b4ac3b45e216a2855e8c0344c504537)  
[pprof profile](https://pprof.me/41ad36edf7f59d9a62085525eba5a51b/?profileType=profile%253Acpu%253Ananoseconds%253Acpu%253Ananoseconds%253Adelta&color_by=filename)

WMI On:  
[config file](https://gist.github.com/braydonk/a86c1c0277a1e29e1bec7f8b0ec60850)  
[pprof profile](https://pprof.me/182f84499e8d35c3f44c3cbb4ae4b9f9/?profileType=profile%253Acpu%253Ananoseconds%253Acpu%253Ananoseconds%253Adelta&color_by=filename)

Important note for comparing the two profiles is that the profile for WMI On is 450ms compared to the WMI Off profile at 2.5s.

---

>  Internally, gopsutils, uses EnumProcesses to get the list of PIDs, then one by one we execute code to get other information about each process. So it calls CreateToolhelp32Snapshot for each process to get the parent PID, instead it should call CreateToolhelp32Snapshot instead of EnumProcesses when obtaining the initial list of PIDs.

I think I'm missing something here; is it possible to call `CreateToolhelp32Snapshot` on a bulk set of processes? The API documentation makes it look like this can only be done for one process at a time: https://learn.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot

> Besides performance the current code has issues regarding correctness and availability of information (processes can terminate between the calls and PIDs can be reused).

This is an issue I've known about for some time but been unsure how to tackle without rearchitecting the scraper. The fact that the list of processes is gathered and then information is fetched individually for each makes it so this problem will always exist on any platform. Much of the scraper was written before my time as a CODEOWNER, but staying relatively the same cross-platform by leaning on `gopsutil` appears to be a goal. It ends up being to the detriment of any platform in the end, but is much simpler to maintain.



I think trying to make WMI fetch all information on Windows would make it so the scraper looks completely different and would diminish that benefit of ease of maintenance (arguably the usage of WMI as it exists and in this PR is already a blow to maintainability, but is necessitated by unacceptable performance on Windows in the current implementation).

---

>  We could do the Win32 API route and in that case some of the info may not be available (parent PID can be made cheaply available without WMI, not sure about handle count).

Had to dig through the archives for this, but back when I introduced the `process.handles` metric in the first place I had an initial version that [used `NtQuerySystemInformation` instead](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22813/commits/fac458d2616b34bbdb3a16d1657ea3db4cae2503#diff-8999f2f1b6a597cd241cd2c692e39d6bfd0498342e94d8d8e85457f92c867304R68). It was rejected in favour of using WMI instead. I'm not sure if the original reservations that were stated at the time still apply. One thing's for sure, it's harder to understand for a general maintainer who may not know how it works. `NtQuerySystemInformation` also doesn't provide a stability guarantee and that was a concern. To my knowledge, the only other way to get Parent PID other than `CreateToolhelp32Snapshot` is `NtQuerySystemInformation` (which I think is what WMI does under the hood?)

Personally I am interested in finding an avenue to open the floor for making this receiver more efficient even if it requires less reliance on `gopsutil` in favour of direct system API usage to allow for gathering the data more efficiently. I'm not sure how to broach the topic more broadly.

---

All other comments on the PR are noted but I won't have free time for a couple days to address them. I plan to ASAP. Thank you for your review!

pjanotti commented:
Thanks for the updated chart and profiles @braydonk!

> > Internally, gopsutils, uses EnumProcesses to get the list of PIDs, then one by one we execute code to get other information about each process. So it calls CreateToolhelp32Snapshot for each process to get the parent PID, instead it should call CreateToolhelp32Snapshot instead of EnumProcesses when obtaining the initial list of PIDs.

> I think I'm missing something here; is it possible to call CreateToolhelp32Snapshot on a bulk set of processes? The API documentation makes it look like this can only be done for one process at a time: https://learn.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot

To get parent PID current code calls gopsutils [`PpidWithContext`](https://github.com/shirou/gopsutil/blob/90e59961ca1cd462850ab7e3345bedc9b9559f4f/process/process_windows.go#L305). Until the `ppid` is cached it needs to call [`getFromSnapProcess`](https://github.com/shirou/gopsutil/blob/90e59961ca1cd462850ab7e3345bedc9b9559f4f/process/process_windows.go#L847) which makes the call to `CreateToolhelp32Snapshot`, but, it must use the flag `TH32CS_SNAPPROCESS` to capture the `ppid` and that takes a snapshot of all processes in the box, the parameter with the PID is ignored in this case. Notice use of win32 APIs `Process32First` and `Process32Next` at https://github.com/shirou/gopsutil/blob/v4.25.1/process/process_windows.go#L847-L868. So if we used this as the mechanism to capture the list of PIDs we would already be getting the `ppid` for each process and in a snapshot that should be consistent. There will be other information that would require us to open each process handle, but, the cost to get the `ppid` would be great reduced.

> staying relatively the same cross-platform by leaning on gopsutil appears to be a goal. It ends up being to the detriment of any platform in the end, but is much simpler to maintain.

I think we can isolate enough the high-level of the receiver from each platform implementation. Perhaps the main thing here is that not many developers in the community know and use Windows so having something like `gopsutils` shielding them from Windows is a more practical reason. That said I'm here to improve the collector experience on Windows 😄 so I can help here.

> Perhaps this is worth addressing at a different angle, such as an entirely new Windows specific process scraper. I'm not sure where that would live or how we'd go about making that though.

At this point I don't think there is enough difference on the data output to justify that. It seems to me that if we make the calls to win32 API directly we can keep most of the non-Windows code as it is. That said it is possible that I'm being too optimistic.

rustyautopsy commented:
We are keeping a close eye on this PR as it would really help us out. We are heavy Windows shop and are actively working with Coralogix to ramp up our Observability efforts. The alternative is running likely running `otel` with `windows_exporter` which is less then ideal.

If there is anyway we can help please ask.

slernercx commented:
@pjanotti steve lerner here- this is one of our customers above who would love to get this merged- they have large fleets of new windows servers and are going 100% otel... obrigadão irmão tava no Rio a semana pasada!

pjanotti commented:
> I think we can isolate enough the high-level of the receiver from each platform implementation.

Here is one proposal on how to do it.

The code uses an interface, [`processHandle`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/f9b5841b6e8f018e86b25546d818da1a842c9de4/receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go#L80) that is a subset of the `gopsutil` `process` methods that would be necessary to implement in a new type to keep the high level code the same between the different OSes.

The way to go about this is by making the implementation of [`getProcessHandlesInternal`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/f9b5841b6e8f018e86b25546d818da1a842c9de4/receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go#L153) platform dependent and on Windows enumerating the processes using `CreateToolhelp32Snapshot`. In the same enumeration it should open each process handle only once and capturing all configured information in a single pass. It is a bigger change than this one, but, it should be the final direction of this scraper on Windows.

There may be some of relatively complex methods of `processHandle` to implement, but at worst they can be implemented in a similar manner to the code on `gopsutil`.

pjanotti commented:
@braydonk I did some tests with a small change based on the idea described above. The change for the numbers below uses `CreateToolhelp32Snapshot` to the enumeration of processes and caches the parent PID and the number of threads (defined on the interface but not used yet). The change needs some polishing, but, doesn't require WMI and since the performance is basically the same as current without collecting the parent process ID I think we don't need any configuration for it. It will be great if you could compare it against the change with WMI. Anyway, if you agree with that I propose that we close this PR and I will polish the [draft one](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38589#issue-2915391777) so we can get it merged soon.

pjanotti commented:
@slernercx @rustyautopsy if you could try the bits from #38589 it will be great.

@braydonk let's close this one in favor of #38589?

braydonk commented:
Yep, closing this in favour of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38589

Please link that PR to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32947 if it hasn't been done already.
